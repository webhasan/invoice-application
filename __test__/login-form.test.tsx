import { act, render, screen, waitFor } from '@testing-library/react';
import LoginForm from '../components/auth/login-form';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

describe('LoginForm', () => {
   test('all require field available in the login form', () => {

      render(<LoginForm submitForm={() => {}}/>);

      const emailField = screen.getByRole('textbox', {name: 'Email Address *'});
      const passwordField = screen.getByLabelText('Password *');
      const submitButton = screen.getByRole('button', {name: 'Login'});

      expect(emailField).toBeInTheDocument();
      expect(passwordField).toBeInTheDocument();
      expect(submitButton).toBeInTheDocument();
   });

   test('available all data-test attributes', async () => {
      const {container} = render(<LoginForm submitForm={() => {}}/>);

      const email = container.querySelector('[data-test="email"]');
      const password = container.querySelector('[data-test="email"]');
      const submitButton = container.querySelector('[data-test="submit-login"]');

      expect(password).toBeInTheDocument();
      expect(email).toBeInTheDocument();
      expect(submitButton).toBeInTheDocument();
   });

   test('show input validation for blank fields', async () => {
      render(<LoginForm submitForm={() => {}}/>);
      const submitButton = screen.getByRole('button', {name: 'Login'});
      userEvent.click(submitButton);
      
      const emailValidation = await screen.findByText('Email is required.');
      const passwordValidation = await screen.findByText('Password is required.');

      expect(emailValidation).toBeInTheDocument();
      expect(passwordValidation).toBeInTheDocument();
   });

   test('form and input button disabled after submit value', async () => {
      await act(async () => {
         await render(<LoginForm submitForm={() => {}}/>);
      });
      
      const emailField = screen.getByRole('textbox', {name: 'Email Address *'});
      const passwordField = screen.getByLabelText('Password *');
      const submitButton = screen.getByRole('button', {name: 'Login'});
      
      await userEvent.type(emailField, 'webhasan24@gmail.com');
      await userEvent.type(passwordField, '1234567');
      userEvent.click(submitButton);

      await waitFor(() => expect(screen.getByRole('button', {name: 'Login'})).toBeDisabled());
      await waitFor(() => expect(screen.getByRole('progressbar')).toBeInTheDocument());
   });

   test('Snapshot of login from of initial stage', () => {
      const {container} = render(<LoginForm submitForm={() => {}}/>);
      expect(container).toMatchSnapshot();
   });

});