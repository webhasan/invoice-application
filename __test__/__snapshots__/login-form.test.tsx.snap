// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Snapshot of login from of initial stage 1`] = `
<div>
  <main
    class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-binzgt"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1fkti26-MuiAvatar-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="LockOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
          />
        </svg>
      </div>
      <h1
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Login
      </h1>
      <form
        class="MuiBox-root css-0"
      >
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="email"
            id="email-label"
          >
            Email Address *
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              data-test="email"
              id="email"
              name="email"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Email Address *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="password"
            id="password-label"
          >
            Password *
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              data-test="password"
              id="password"
              name="password"
              type="password"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
            >
              <button
                aria-label="toggle password visibility"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="VisibilityIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Password *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root css-1qudo6r-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
          data-test="submit-login"
          id=":re:"
          tabindex="0"
          type="submit"
        >
          Login
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <div
          class="text-center"
        >
          <a
            class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineAlways css-101ca9i-MuiTypography-root-MuiLink-root"
            href="/sign-up"
          >
            Don't have an account? Sign Up
          </a>
        </div>
      </form>
    </div>
  </main>
</div>
`;
